yo1k/tic_tac_toe/core/game.py

```diff
@@ -0,0 +1,296 @@
+            assert False
+        return f"{type(self)}(" \
+               f"{action})"
+
+
+class ActionQueue(ABC):
+    @abstractmethod
+    def next(self) -> Optional[Action]:
+        pass
+
+
+class Logic:
+    """Game logic."""
+
+    def __init__(self, action_queues: Sequence[ActionQueue]):
+        """Indexes in `action_queues` correspond to indexes in `State.players`."""
+        self.__action_queues = action_queues
+
+    def advance(self, state: State):
+        turn = Logic.__turn(state)
+        action = self.__action_queues[turn].next()
+        if action is None:
+            return None
+        elif action.start is True:
+            assert state.phase is Phase.ONSET or state.phase is Phase.OUTROUND
+            Logic.__start(state)
+        elif action.surrender is True:
+            assert state.phase is Phase.INROUND
+            Logic.__surrender(state)
+        elif action.occupy is not None:
+            assert state.phase is Phase.INROUND
+            Logic.__occupy(state, action.occupy)
+        else:
+            assert False
+
+    @staticmethod
+    def __turn(state: State) -> int:
+        """Returns `turn` - the index of the current `Player` in `State.players`.
+
+        `turn` is calculated in such a way as to alternate players order in different
+        `round` keeping the same `Player` index in the `State.players`.
+        """
+        return (state.step + state.round) % 2
+
+    @staticmethod
+    def __occupy(state: State, cell: Cell):
+        turn = Logic.__turn(state)
+        mark = state.players[turn].mark()
+        if state.board._get(cell) is None:
+            state.board._set(cell, mark)
+        else:
+            pass
+        if Logic._win_condition(state.board, cell):
+            Logic.__win(state)
+        elif Logic._draw_condition(state.step):
+            Logic.__draw(state)
+        else:
+            state.step += 1
+
+    @staticmethod
+    def _win_condition(board: Board, cell: Cell) -> bool:
+        win_condition = False
+        horizontal_match_number = 0
+        vertical_match_number = 0
+        first_diagonal_match_number = 0
+        second_diagonal_match_number = 0
+        x = cell.x
+        y = cell.y
+        mark = board._get(Cell(x, y))
+        for i in range(Board.size()):
+            if board._get(Cell(i, y)) == mark:
+                horizontal_match_number += 1
+            if board._get(Cell(x, i)) == mark:
+                vertical_match_number += 1
+            if board._get((Cell(i, i))) == mark:
+                first_diagonal_match_number += 1
+            if board._get(Cell(i, Board.size() - 1 - i)) == mark:
+                second_diagonal_match_number += 1
+        if horizontal_match_number == Board.size() or \
+           vertical_match_number == Board.size() or \
+           first_diagonal_match_number == Board.size() or \
+           second_diagonal_match_number == Board.size():
+            win_condition = True
+        return win_condition
```

Comment:
Replace with
```python
return h_match == Board.size() or v_match == Board.size() or \
    d1_match == Board.size() or d2_match == Board.size()
```

P.S. If you decide to keep current variable names, then keep them. The idea is that you need neither the `if` statement, nor the `win_condition` variable.

Reply:
Done.